src/main/java/build/buf/protovalidate/CustomOverload.java

```diff
@@ -527,66 +557,81 @@ private static Val uniqueList(Lister list) {
+    if (val.length() > 253) {
       return false;
     }
-    String s = Ascii.toLowerCase(host.endsWith(".") ? host.substring(0, host.length() - 1) : host);
-    Iterable<String> parts = Splitter.on('.').split(s);
+
+    String str;
+    if (val.endsWith(".")) {
+      str = val.substring(0, val.length() - 1);
+    } else {
+      str = val;
+    }
+
     boolean allDigits = false;
+
+    String[] parts = str.toLowerCase(Locale.getDefault()).split("\\.", -1);
+
+    // split hostname on '.' and validate each part
     for (String part : parts) {
       allDigits = true;
-      int l = part.length();
-      if (l == 0 || l > 63 || part.charAt(0) == '-' || part.charAt(l - 1) == '-') {
+
+      // if part is empty, longer than 63 chars, or starts/ends with '-', it is invalid
+      int len = part.length();
+      if (len == 0 || len > 63 || part.startsWith("-") || part.endsWith("-")) {
         return false;
       }
-      for (int i = 0; i < l; i++) {
-        char ch = part.charAt(i);
-        if (!Ascii.isLowerCase(ch) && !isDigit(ch) && ch != '-') {
+
+      // for each character in part
+      for (int i = 0; i < part.length(); i++) {
+        char c = part.charAt(i);
+        // if the character is not a-z, 0-9, or '-', it is invalid
+        if ((c < 'a' || c > 'z') && (c < '0' || c > '9') && c != '-') {
           return false;
         }
-        allDigits = allDigits && isDigit(ch);
+
+        allDigits = allDigits && c >= '0' && c <= '9';
       }
     }
+
     // the last part cannot be all numbers
     return !allDigits;
   }
 
-  private static boolean isDigit(char c) {
-    return c >= '0' && c <= '9';
-  }
-
   /**
-   * Validates if the input string is a valid IP address.
+   * Returns true if the string is an IPv4 or IPv6 address, optionally limited to a specific
+   * version.
+   *
+   * <p>Version 0 means either 4 or 6. Passing a version other than 0, 4, or 6 always returns false.
+   *
+   * <p>IPv4 addresses are expected in the dotted decimal format, for example "192.168.5.21". IPv6
+   * addresses are expected in their text representation, for example "::1", or
+   * "2001:0DB8:ABCD:0012::0".
    *
-   * @param addr The input string to validate as an IP address.
-   * @param ver The IP version to validate against (0 for any version, 4 for IPv4, 6 for IPv6).
-   * @return {@code true} if the input string is a valid IP address of the specified version, {@code
-   *     false} otherwise.
+   * <p>Both formats are well-defined in the internet standard RFC 3986. Zone identifiers for IPv6
+   * addresses (for example "fe80::a%en1") are supported.
    */
-  private static boolean validateIP(String addr, long ver) {
-    InetAddress address;
-    try {
-      address = InetAddresses.forString(addr);
-    } catch (Exception e) {
-      return false;
-    }
-    if (ver == 0L) {
-      return true;
+  private static boolean isIP(String addr, long ver) {
+    if (ver == 6L) {
+      return new Ipv6(addr).address();
```

Comment:
This is not very idiomatic Java here - having to create an empty object then calling a method on it to determine if it is valid.

I think we should make these functions like `isIPv4` and `isIPv6` that minimize allocations and state if possible.

Reply:
OK, this is where the nuance of trying to stay consistent comes into play. This allocation is done across impls, so should we refactor those also? cc @timostamm